distributed/worker.py

```diff
@@ -920,3 +717,666 @@ def run(worker, stream, function=None, args=(), kwargs={}):
+                assert key not in self.heap
+
+            del self.waiting_for_data[key]
+            heapq.heappush(self.heap, (self.priorities[key], key))
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    def transition_ready_executing(self, key):
+        try:
+            if self.validate:
+                assert key not in self.waiting_for_data
+                # assert key not in self.data
+                assert self.task_state[key] == 'ready'
+                assert key not in self.heap
+                assert all(dep in self.data for dep in self.dependencies[key])
+
+            self.executing.add(key)
+            self.loop.add_callback(self.execute, key)
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    def transition_executing_done(self, key):
+        try:
+            if self.validate:
+                assert key in self.executing or key in self.long_running
+                assert key not in self.waiting_for_data
+                assert key not in self.heap
+
+            if self.task_state[key] == 'executing':
+                self.executing.remove(key)
+            elif self.task_state[key] == 'long-running':
+                self.long_running.remove(key)
+            if self.batched_stream:
+                self.batched_stream.send(self.response[key])
+
+        except StreamClosedError:
+            logger.info("Stream closed")
+            self._close(report=False)
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    def transition_executing_long_running(self, key):
+        try:
+            if self.validate:
+                assert key in self.executing
+
+            self.executing.remove(key)
+            self.long_running.add(key)
+
+            self.ensure_computing()
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    ##########################
+    # Gather Data from Peers #
+    ##########################
+
+    def ensure_communicating(self):
+        try:
+            while self.data_needed and len(self.connections) < self.total_connections:
+                logger.debug("Ensure communicating.  Pending: %d.  Connections: %d/%d",
+                             len(self.data_needed),
+                             len(self.connections),
+                             self.total_connections)
+
+                key = self.data_needed[0]
+
+                if key not in self.tasks:
+                    self.data_needed.popleft()
+                    continue
+
+                if self.task_state.get(key) != 'waiting':
+                    self.log.append((key, 'communication pass'))
```

Comment:
1.  If the key was removed by some other event `key not in self.task_state` but wasn't removed from `data_needed` (removal from the deque is expensive)
2.  If they key was jumped to some other state, such as `'memory'` if we gather the key as part of a dependency for  some other task

Neither of these occur during normal operation